pkg/azuredisk/device_perf_optimization_linux.go

```diff
@@ -0,0 +1,296 @@
+// +build linux
+
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package azuredisk
+
+import (
+	"fmt"
+	"math"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strconv"
+	"strings"
+
+	"k8s.io/klog/v2"
+)
+
+type DeviceHelper struct{}
+
+func (deviceHelper *DeviceHelper) OptimizeDevicePerformance(nodeInfo *NodeInfo, diskSkus map[string]map[string]DiskSkuInfo, devicePath string, attributes map[string]string) (err error) {
+
+	klog.V(2).Infof("OptimizeDevicePerformance: Tuning settings for %s", devicePath)
+
+	if nodeInfo == nil {
+		err = fmt.Errorf("Node info is not provided. Error: invalid parameter")
+		klog.Errorf("Error: %s", err)
+		return err
+	}
+
+	if diskSkus == nil {
+		err = fmt.Errorf("Disk SKUs are not provided. Error: invalid parameter")
+		klog.Errorf("Error: %s", err)
+		return err
+
+	}
+
+	mode, profile, accountType, diskSizeGibStr, diskIopsStr, diskBwMbpsStr := GetDiskPerfAttributes(attributes)
+
+	if !IsPerfTuningEnabled(mode) {
+		klog.Warningf("OptimizeDevicePerformance: Perf tuning not enabled for mode %s devicepath %s", mode, devicePath)
+		return nil
+	}
+
+	if !AccountSupportsPerfOptimization(accountType) {
+		klog.Warningf("OptimizeDevicePerformance: Perf tuning not enabled for accountType %s devicepath %s", accountType, devicePath)
+		return nil
+	}
+
+	queueDepth, nrRequests, scheduler, maxSectorsKb, readAheadKb, err := getOptimalDeviceSettings(nodeInfo, diskSkus, mode, profile, accountType, diskSizeGibStr, diskIopsStr, diskBwMbpsStr)
+	if err != nil {
+		klog.Errorf("OptimizeDevicePerformance: Failed to get optimal settings for mode %s profile %s accountType %s device %s Error: %s", mode, profile, accountType, devicePath, err)
+		return err
+	}
+
+	deviceName, err := getDeviceName(devicePath)
+	if err != nil {
+		klog.Errorf("OptimizeDevicePerformance: Could not get deviceName for %s. Error: %s", devicePath, err)
+		return err
+	}
+
+	klog.V(2).Infof("OptimizeDevicePerformance: Tuning settings for devicePath %s, deviceName %s, profile %s queueDepth %s nrRequests %s scheduler %s maxSectorsKb %s readAheadKb %s",
+		devicePath,
+		deviceName,
+		profile,
+		queueDepth,
+		nrRequests,
+		scheduler,
+		maxSectorsKb,
+		readAheadKb)
+
+	err = echoToFile(maxSectorsKb, filepath.Join("/sys/block", deviceName, "queue/max_sectors_kb"))
+	if err != nil {
+		klog.Errorf("OptimizeDevicePerformance: Could not set max_sectors_kb for device %s. Error: %v", deviceName, err)
```

Comment:
no need to log every errors in the function, just log outsider

Reply:
Don't we lose the error stack in the kind of logging you're suggesting? The regimen you're suggesting would result in one log at a top level function and if the error is generic enough, it would be hard to track the real root cause, line number etc of the failure. It probably makes sense to do it for smaller functions with fewer places throwing errors.

For example in this case If I don't format and log the error thrown by the exact setting, I would not know which setting actually failed. Making it harder for me to debug.